torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2662,4 +2692,98 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
   return work;
 }
 
+void ProcessGroupGloo::monitoredBarrier(const BarrierOptions& opts) {
+  // Use default timeout if no timeout was specified.
+  auto monitoredBarrierTimeout =
+      (opts.timeout == kUnsetTimeout) ? this->options_->timeout : opts.timeout;
+  auto rank = this->getRank();
+  auto t1 = nextTag();
+  auto t2 = nextTag();
+  std::vector<at::Tensor> commTensor = {at::tensor({rank})};
+  // only enforce timeout on rank 0. This is so that other ranks aren't timed
+  // out first, bringing down the job without reporting which rank timed out.
+  if (rank != 0) {
+    auto sendWork = send(commTensor, 0, t1);
+    auto recvWork = recv(commTensor, 0, t2);
+    sendWork->wait();
+    recvWork->wait();
+    return;
+  }
+  auto startTime = std::chrono::steady_clock::now();
+  auto worldSize = this->getSize();
+  // Holds mapping of rank to pair of recv/send work.
+  std::unordered_map<
+      int,
+      std::tuple<
+          c10::intrusive_ptr<ProcessGroup::Work>,
+          c10::intrusive_ptr<ProcessGroup::Work>>>
+      works;
+  // Kick off all work
+  for (int dstRank = 1; dstRank < worldSize; ++dstRank) {
+    auto recvWork = recv(commTensor, dstRank, t1);
+    auto sendWork = send(commTensor, dstRank, t2);
+    works.insert(
+        {dstRank, std::make_tuple(std::move(recvWork), std::move(sendWork))});
+  }
+
+  const std::string kNoRemainingTimeError = c10::str(
+      "Rank ",
+      rank,
+      " timed out in monitoredBarrier after ",
+      monitoredBarrierTimeout.count(),
+      " ms.");
+  std::vector<int> processedRanks;
+
+  // Wait for send/recv from all ranks
+  for (auto& work : works) {
+    for (int i = 0; i < 2; ++i) {
+      // Compute the remaining time for every call to wait() instead of using
+      // the whole timeout every call. Error out if no time remains.
+      auto remainingTime = getRemainingTime(startTime, monitoredBarrierTimeout);
+      if (remainingTime.count() < 0) {
+        std::string rankInfo;
+        if (processedRanks.size() > 0) {
+          rankInfo = c10::str(
+              "Successfully processed ranks: ",
+              c10::Join(", ", processedRanks));
```

Comment:
should not it also include this rank info and saying current rank=work.first did not pass the monitored barrier?

Reply:
The rank logging this information is always rank 0, and I'm not sure if we can say current rank = work.first did not pass the monitored barrier.

In this codepath, we basically ran out of time during processing and did not call into `work.wait()`. The rank may indeed have been healthy but we ran out of time to process it, so we cannot surely say either way and we need to fail here due to the timeout.

In the general case, we fail in `work.wait()` where our timeout is exhausted and that is handled below, where we do log the failed rank.

Reply:
I see, that makes sense!